keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,205 @@
+
+Starting in 1.16 the legacy code block should be gated on `LegacyNodeRoleBehavior=on`
+
+
+#### Node controller excludes master nodes from consideration for eviction
+
+The `k8s.io/kubernetes/pkg/util/system/IsMasterNode(nodeName)` function is used by the NodeLifecycleController to exclude nodes with a node name that ends in `master` or starts with `master-` when considering whether to mark nodes as disrupted. A recent PR attempted to change this to use node-roles and was blocked. Instead, the controller should be updated to use a label `node-controller.kubernetes.io/exclude-disruption` to decide whether to exclude nodes from being considered for disruption handling.
+
+
+#### Kubernetes e2e tests
+
+The e2e tests use a number of heuristics including the `IsMasterNode(nodeName)` function and the node-roles labels to select nodes. In order for conformant Kubernetes clusters to run the tests, the e2e suite must change to use individual user-provided label selectors to identify nodes to test, nodes that have special rules for testing unusual cases, and for other selection behaviors. The label selectors may be defaulted by the test code to their current values, as long as a conformant cluster operator can execute the e2e suite against an arbitrary cluster.
+
+QUESTION: Is a single label selector sufficient to identify nodes to test?
+
+
+## Design Details
+
+### Test Plan
+
+* Unit tests to verify selection using feature gates
+
+### Graduation Criteria
+
+* New labels and feature flags become beta after one release, GA and defaulted on after two, and are removed after 4 releases.
+* Documentation for migrating to the new labels is available in 1.17.
+
+### Upgrade / Downgrade Strategy
+
+As described in the migration process, deployers and administrators have 2 releases to migrate their clusters.
+
+### Version Skew Strategy
+
+Controllers are updated after the control plane, so consumers must update the labels on their nodes before they update controller processes in 1.18.
+
+## Implementation History
+
+- 2019-07-16: Created
+
+## Alternatives
+
+Allowing core components to use node-role labels was considered and rejected because:
+
+1. The number of impacted components and clusters is anticipated to be small
+2. There are 4 releases to update within
+3. Some deployment topologies (self-hosted clusters) cannot use service load balancers for control plane components today
+4. We wish to prevent additional confusion by providing clear guidance.
+
+
+## Reference
+
+* https://github.com/kubernetes/kubernetes/pull/35975
+* https://github.com/kubernetes/kubernetes/pull/39112
+* https://github.com/kubernetes/kubernetes/pull/76654
+* https://github.com/kubernetes/kubernetes/pull/80021
```

Comment:
https://groups.google.com/d/msg/kubernetes-sig-architecture/ZKUOPy2PNJ4/lDh4hs4HBQAJ is a useful thread to reference